### ISSUE TITLE
Provider produced inconsistent result after apply — sometimes

### ISSUE THREAD

**Issue by alexs77** [iLert/terraform-provider-ilert#15]
I'm running `env TF_LOG=DEBUG TF_LOG_PATH=/tmp/$(pwd | tr / _).$(date | tr ' ' _).apply-auto-approve.log terraform apply -auto-approve -var-file=environments/devops_shr_testalert.tfvars`. It *sometimes* fails like so:

```sh
module.err_vm_cpu.ilert_connector.webhooks: Refreshing state... [id=…-3ba3dc6b5e76]
module.err_vm_cpu.ilert_connector.emails: Refreshing state... [id=…-b4b068bdb3ce]
module.warn_vm_cpu.ilert_connector.webhooks: Refreshing state... [id=…-aa9fdc039771]
module.warn_vm_cpu.ilert_connector.emails: Refreshing state... [id=…-3a96b0fe25b0]
module.warn_vm_cpu.ilert_connector.msteams["https://adv.webhook.office.com/webhookb2/…/IncomingWebhook/0de243df23da43448f8114ff543aedd6/9f48f9a9-3efe-46fd-a9de-faacf7dafd2b"]: Refreshing state... [id=…-6d941d8afd1a]
module.err_vm_cpu.ilert_connector.msteams["https://adv.webhook.office.com/webhookb2/…/IncomingWebhook/5788d91f3040432e84588cd923054b38/9f48f9a9-3efe-46fd-a9de-faacf7dafd2b"]: Refreshing state... [id=…-8e960bcdca76]
module.err_vm_cpu.ilert_alert_source.this: Refreshing state... [id=2225345]
module.warn_vm_cpu.ilert_alert_source.this: Refreshing state... [id=2225409]
module.err_vm_cpu.ilert_connection.msteams["https://adv.webhook.office.com/webhookb2/…/IncomingWebhook/5788d91f3040432e84588cd923054b38/9f48f9a9-3efe-46fd-a9de-faacf7dafd2b"]: Refreshing state... [id=…-21df65ed562a]
module.err_vm_cpu.ilert_connection.emails["john.doe@example.com"]: Refreshing state... [id=…-c3180d1b9bd1]
module.err_vm_cpu.ilert_connection.emails["mary.major@example.net"]: Refreshing state... [id=…-2fd0bf295a4e]
module.err_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"]: Refreshing state... [id=…-d8b157664e73]
module.warn_vm_cpu.ilert_connection.msteams["https://adv.webhook.office.com/webhookb2/…/IncomingWebhook/0de243df23da43448f8114ff543aedd6/9f48f9a9-3efe-46fd-a9de-faacf7dafd2b"]: Refreshing state... [id=…-da50290dd929]
module.warn_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"]: Refreshing state... [id=…-d12e0fc614fc]
module.warn_vm_cpu.ilert_connection.emails["mary.major@example.net"]: Refreshing state... [id=…-59c9d5a8ca53]
module.warn_vm_cpu.ilert_connection.emails["john.doe@example.com"]: Refreshing state... [id=…-188559615816]
module.err_vm_cpu.azurerm_monitor_action_group.this: Refreshing state... [id=/subscriptions/…-4fc7babf5891/resourceGroups/DevOpsTestAlert/providers/Microsoft.Insights/actionGroups/🔴 (E) devops-shr-es-vm-testalert VM CPU Alert]
azurerm_log_analytics_workspace.this[0]: Refreshing state... [id=/subscriptions/…-4fc7babf5891/resourceGroups/TpxPrd/providers/Microsoft.OperationalInsights/workspaces/devops-shr-testalert-law-alert]
module.warn_vm_cpu.azurerm_monitor_action_group.this: Refreshing state... [id=/subscriptions/…-4fc7babf5891/resourceGroups/DevOpsTestAlert/providers/Microsoft.Insights/actionGroups/🟡 (W) devops-shr-es-vm-testalert VM CPU Alert]
azurerm_application_insights.this[0]: Refreshing state... [id=/subscriptions/…-4fc7babf5891/resourceGroups/TpxPrd/providers/Microsoft.Insights/components/tpx-shr-alert]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # module.err_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"] must be replaced
-/+ resource "ilert_connection" "emails" {
      ~ created_at    = "2022-04-14T06:25:13.175Z" -> (known after apply)
      ~ id            = "…-d8b157664e73" -> (known after apply)
        name          = "Emails 6f687034-f70e-5e5b-befb-df3729d6152e Connection - 🔴 (E) devops-shr-es-vm-testalert VM CPU Alert"
      ~ updated_at    = "2022-04-14T06:25:13.175Z" -> (known after apply)
        # (2 unchanged attributes hidden)


      ~ connector {
          + id   = "…-b4b068bdb3ce"
          + type = "email"
        }

      - email { # forces replacement
          - recipients = [] -> null
        }
        # (1 unchanged block hidden)
    }

  # module.err_vm_cpu.ilert_connection.emails["john.doe@example.com"] must be replaced
-/+ resource "ilert_connection" "emails" {
      ~ created_at    = "2022-04-14T06:25:13.204Z" -> (known after apply)
      ~ id            = "…-c3180d1b9bd1" -> (known after apply)
        name          = "Emails 14e6862a-eca5-506f-a5f0-91eeed852f9e Connection - 🔴 (E) devops-shr-es-vm-testalert VM CPU Alert"
      ~ updated_at    = "2022-04-14T06:25:13.204Z" -> (known after apply)
        # (2 unchanged attributes hidden)


      ~ connector {
          + id   = "…-b4b068bdb3ce"
          + type = "email"
        }

      - email { # forces replacement
          - recipients = [] -> null
        }
        # (1 unchanged block hidden)
    }

  # module.err_vm_cpu.ilert_connection.emails["mary.major@example.net"] must be replaced
-/+ resource "ilert_connection" "emails" {
      ~ created_at    = "2022-04-14T06:25:13.070Z" -> (known after apply)
      ~ id            = "…-2fd0bf295a4e" -> (known after apply)
        name          = "Emails 1ca283fb-4355-54ab-abc3-8136cf9e8ac7 Connection - 🔴 (E) devops-shr-es-vm-testalert VM CPU Alert"
      ~ updated_at    = "2022-04-14T06:25:13.070Z" -> (known after apply)
        # (2 unchanged attributes hidden)


      ~ connector {
          + id   = "…-b4b068bdb3ce"
          + type = "email"
        }

      - email { # forces replacement
          - recipients = [] -> null
        }
        # (1 unchanged block hidden)
    }

  # module.warn_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"] must be replaced
-/+ resource "ilert_connection" "emails" {
      ~ created_at    = "2022-04-14T06:25:13.159Z" -> (known after apply)
      ~ id            = "…-d12e0fc614fc" -> (known after apply)
        name          = "Emails 6f687034-f70e-5e5b-befb-df3729d6152e Connection - 🟡 (W) devops-shr-es-vm-testalert VM CPU Alert"
      ~ updated_at    = "2022-04-14T06:25:13.159Z" -> (known after apply)
        # (2 unchanged attributes hidden)


      ~ connector {
          + id   = "…-3a96b0fe25b0"
          + type = "email"
        }

      - email { # forces replacement
          - recipients = [] -> null
        }
        # (1 unchanged block hidden)
    }

  # module.warn_vm_cpu.ilert_connection.emails["john.doe@example.com"] must be replaced
-/+ resource "ilert_connection" "emails" {
      ~ created_at    = "2022-04-14T06:25:13.197Z" -> (known after apply)
      ~ id            = "…-188559615816" -> (known after apply)
        name          = "Emails 14e6862a-eca5-506f-a5f0-91eeed852f9e Connection - 🟡 (W) devops-shr-es-vm-testalert VM CPU Alert"
      ~ updated_at    = "2022-04-14T06:25:13.197Z" -> (known after apply)
        # (2 unchanged attributes hidden)


      ~ connector {
          + id   = "…-3a96b0fe25b0"
          + type = "email"
        }

      - email { # forces replacement
          - recipients = [] -> null
        }
        # (1 unchanged block hidden)
    }

  # module.warn_vm_cpu.ilert_connection.emails["mary.major@example.net"] must be replaced
-/+ resource "ilert_connection" "emails" {
      ~ created_at    = "2022-04-14T06:25:13.081Z" -> (known after apply)
      ~ id            = "…-59c9d5a8ca53" -> (known after apply)
        name          = "Emails 1ca283fb-4355-54ab-abc3-8136cf9e8ac7 Connection - 🟡 (W) devops-shr-es-vm-testalert VM CPU Alert"
      ~ updated_at    = "2022-04-14T06:25:13.081Z" -> (known after apply)
        # (2 unchanged attributes hidden)


      ~ connector {
          + id   = "…-3a96b0fe25b0"
          + type = "email"
        }

      - email { # forces replacement
          - recipients = [] -> null
        }
        # (1 unchanged block hidden)
    }

Plan: 6 to add, 0 to change, 6 to destroy.
module.warn_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"]: Destroying... [id=…-d12e0fc614fc]
module.warn_vm_cpu.ilert_connection.emails["mary.major@example.net"]: Destroying... [id=…-59c9d5a8ca53]
module.err_vm_cpu.ilert_connection.emails["john.doe@example.com"]: Destroying... [id=…-c3180d1b9bd1]
module.err_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"]: Destroying... [id=…-d8b157664e73]
module.err_vm_cpu.ilert_connection.emails["mary.major@example.net"]: Destroying... [id=…-2fd0bf295a4e]
module.warn_vm_cpu.ilert_connection.emails["john.doe@example.com"]: Destroying... [id=…-188559615816]
module.warn_vm_cpu.ilert_connection.emails["mary.major@example.net"]: Destruction complete after 0s
module.warn_vm_cpu.ilert_connection.emails["mary.major@example.net"]: Creating...
module.warn_vm_cpu.ilert_connection.emails["john.doe@example.com"]: Destruction complete after 0s
module.warn_vm_cpu.ilert_connection.emails["john.doe@example.com"]: Creating...
module.err_vm_cpu.ilert_connection.emails["john.doe@example.com"]: Destruction complete after 0s
module.err_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"]: Destruction complete after 0s
module.err_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"]: Creating...
module.err_vm_cpu.ilert_connection.emails["john.doe@example.com"]: Creating...
module.warn_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"]: Destruction complete after 0s
module.err_vm_cpu.ilert_connection.emails["mary.major@example.net"]: Destruction complete after 0s
module.warn_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"]: Creating...
module.err_vm_cpu.ilert_connection.emails["mary.major@example.net"]: Creating...
module.warn_vm_cpu.ilert_connection.emails["mary.major@example.net"]: Creation complete after 0s [id=ff2f1d59-3ea6-448b-85fa-05fd3be940cd]
module.warn_vm_cpu.ilert_connection.emails["john.doe@example.com"]: Creation complete after 0s [id=b1cdd4c7-c4eb-4dcf-92ec-ca5417e9562f]
module.warn_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"]: Creation complete after 0s [id=e9d72041-ebc0-452c-8a5a-7da3f1bf8273]
module.err_vm_cpu.ilert_connection.emails["mary.major@example.net"]: Creation complete after 0s [id=c4f6efc3-8da7-4019-893d-56b2b8784a5a]
╷
│ Warning: Experimental feature "module_variable_optional_attrs" is active
│ 
│   on providers.tf line 45, in terraform:
│   45:   experiments = [module_variable_optional_attrs]
│ 
│ Experimental features are subject to breaking changes in future minor or patch releases, based on feedback.
│ 
│ If you have feedback on the design of this feature, please open a GitHub issue to discuss it.
╵
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to module.err_vm_cpu.ilert_connection.emails["fred.bloggs@example.org"], provider "provider[\"registry.terraform.io/ilert/ilert\"]" produced an unexpected new value: Root resource
│ was present, but now absent.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to module.err_vm_cpu.ilert_connection.emails["john.doe@example.com"], provider "provider[\"registry.terraform.io/ilert/ilert\"]" produced an unexpected new value: Root resource was
│ present, but now absent.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
```

It does *NOT* **ALWAYS** fail with this. Doing experiments, I noticed that it is more likely to fail, the more "`ilert_connector`" and/or "`ilert_connection`" resources are to be created in one `terraform apply` run.

In the attached source code, this can easily be done by adding a few more of these blocks:

```hcl
module "err_vm_cpu" {
  source = "./modules/ilert_action"
  name                = "${local.category["err"]["badge"]} ${var.vm["name"]} VM CPU Alert"
  nm                  = var.nm["err"]["vm"]["cpu"]
  resource_group_name = var.resource_group_name
  short_name          = "vm_cpu_err"
}
```

- Terraform debug log ⇒ [_home_askwar_Code_Sharedien.DevOps.Alerting.Sharedien_Terraform.Debug.Do_14_Apr_2022_08:25:13_CEST.apply-auto-approve.log](https://github.com/iLert/terraform-provider-ilert/files/8487125/_home_askwar_Code_Sharedien.DevOps.Alerting.Sharedien_Terraform.Debug.Do_14_Apr_2022_08.25.13_CEST.apply-auto-approve.log)
- Source code ⇒ [terraform-ilert-inconsistent-result.tar.gz](https://github.com/iLert/terraform-provider-ilert/files/8487129/terraform-ilert-inconsistent-result.tar.gz)

**Comment by yacut** [comment]
@alexs77 Thanks for the bug report. I will look into it right now as well as other open issues you have created.

**Comment by STLVRTX** [comment]
Similar to #16 and #17, this issue could not be reproduced anymore and most likely fixed in version >1.6, closing.
